File: frontend/src/types/cost-history.ts
/**
 * Types for historical cost tracking (US0183).
 *
 * EP0005: Cost Tracking - Historical Cost Tracking
 */

/** Single cost history record */
export interface CostHistoryItem {
  /** ISO date string or period label (e.g., '2026-01', 'W01') */
  date: string;
  /** Estimated kWh consumed */
  estimated_kwh: number;
  /** Estimated cost in configured currency */
  estimated_cost: number;
  /** Electricity rate per kWh */
  electricity_rate: number;
  /** Server ID (if filtered) */
  server_id?: string;
  /** Server hostname (if available) */
  server_hostname?: string;
}

/** Aggregation level for cost history queries */
export type AggregationType = 'daily' | 'weekly' | 'monthly';

/** Response from GET /costs/history */
export interface CostHistoryResponse {
  /** Cost history records */
  items: CostHistoryItem[];
  /** Aggregation level used */
  aggregation: AggregationType;
  /** Start of date range */
  start_date: string;
  /** End of date range */
  end_date: string;
  /** Currency symbol */
  currency_symbol: string;
}

/** Monthly cost summary record */
export interface MonthlySummaryItem {
  /** Year-month string 'YYYY-MM' */
  year_month: string;
  /** Total cost for the month */
  total_cost: number;
  /** Total kWh for the month */
  total_kwh: number;
  /** Previous month's cost (null for first month) */
  previous_month_cost: number | null;
  /** Month-over-month change percentage */
  change_percent: number | null;
}

/** Response from GET /costs/summary/monthly */
export interface MonthlySummaryResponse {
  /** Monthly cost summaries */
  months: MonthlySummaryItem[];
  /** Year of the summary */
  year: number;
  /** Year-to-date total cost */
  year_to_date_cost: number;
  /** Currency symbol */
  currency_symbol: string;
}

/** Period options for server cost history */
export type CostHistoryPeriod = '7d' | '30d' | '90d' | '12m';

/** Response from GET /servers/{id}/costs/history */
export interface ServerCostHistoryResponse {
  /** Server identifier */
  server_id: string;
  /** Server hostname */
  hostname: string;
  /** Period used for the query */
  period: CostHistoryPeriod;
  /** Cost history records */
  items: CostHistoryItem[];
  /** Currency symbol */
  currency_symbol: string;
}

File: backend/src/homelab_cmd/api/schemas/cost_history.py
"""Pydantic schemas for Cost History API endpoints.

US0183: Historical Cost Tracking (EP0005)

Defines schemas for cost history, monthly summary, and per-server history APIs.
"""

from pydantic import BaseModel, Field


class CostHistoryItem(BaseModel):
    """Single cost history record."""

    date: str = Field(description="ISO date string or period label (e.g., '2026-01', 'W01')")
    estimated_kwh: float = Field(description="Estimated kWh consumed")
    estimated_cost: float = Field(description="Estimated cost in configured currency")
    electricity_rate: float = Field(description="Electricity rate per kWh")
    server_id: str | None = Field(default=None, description="Server ID (if filtered)")
    server_hostname: str | None = Field(default=None, description="Server hostname (if available)")


class CostHistoryResponse(BaseModel):
    """Response for cost history endpoint.

    AC2: Historical cost API with filtering and aggregation.
    """

    items: list[CostHistoryItem] = Field(description="Cost history records")
    aggregation: str = Field(description="Aggregation level: 'daily', 'weekly', or 'monthly'")
    start_date: str = Field(description="Start of date range (ISO format)")
    end_date: str = Field(description="End of date range (ISO format)")
    currency_symbol: str = Field(description="Currency symbol (e.g., '$', '£')")


class MonthlySummaryItem(BaseModel):
    """Monthly cost summary record."""

    year_month: str = Field(description="Year-month string 'YYYY-MM'")
    total_cost: float = Field(description="Total cost for the month")
    total_kwh: float = Field(description="Total kWh for the month")
    previous_month_cost: float | None = Field(
        default=None, description="Previous month's cost (for first month, None)"
    )
    change_percent: float | None = Field(
        default=None, description="Month-over-month change percentage"
    )


class MonthlySummaryResponse(BaseModel):
    """Response for monthly summary endpoint.

    AC5: Monthly cost summary with YTD.
    """

    months: list[MonthlySummaryItem] = Field(description="Monthly cost summaries")
    year: int = Field(description="Year of the summary")
    year_to_date_cost: float = Field(description="Year-to-date total cost")
    currency_symbol: str = Field(description="Currency symbol (e.g., '$', '£')")


class ServerCostHistoryResponse(BaseModel):
    """Response for per-server cost history endpoint.

    AC4: Per-server cost history.
    """

    server_id: str = Field(description="Server identifier")
    hostname: str = Field(description="Server hostname")
    period: str = Field(description="Period: '7d', '30d', '90d', or '12m'")
    items: list[CostHistoryItem] = Field(description="Cost history records")
    currency_symbol: str = Field(description="Currency symbol (e.g., '$', '£')")
